File: simple_qrs.js
const fs = require("fs-extra");
const QRCode = require("qr-image");
const Jimp = require("jimp");

const NUM_OF_QRS = 10000; // Change this to the desired number of QR codes
const OUTPUT_FOLDER = "qr_codes";
const WEBSITE_URL = "https://www.albertoqrtest.com/QR_";
const PRINT_STATS_INTERVAL = NUM_OF_QRS / 10;

const generateQRCode = async (qrId) => {
  return new Promise((resolve, reject) => {
    const qrUrl = `${WEBSITE_URL}${qrId}`;
    const qrImage = QRCode.image(qrUrl, { type: "png", size: 10 });

    const qrImagePath = `${OUTPUT_FOLDER}/QR_${qrId}.png`;

    qrImage.pipe(fs.createWriteStream(qrImagePath));
    qrImage.on("end", () => {
      resolve(qrImagePath);
    });
    qrImage.on("error", (error) => {
      reject(error);
    });
  });
};

const calculateStats = () => {
  const files = fs.readdirSync(OUTPUT_FOLDER);
  const totalSize = files.reduce(
    (acc, file) => acc + fs.statSync(`${OUTPUT_FOLDER}/${file}`).size,
    0
  );
  const averageSize = totalSize / files.length;

  console.log("------ Stats ------");
  console.log(`Total number of QR codes: ${NUM_OF_QRS}`);
  console.log(`Total folder size: ${totalSize} bytes`);
  console.log(`Average QR code size: ${averageSize} bytes`);
};

const generateQRs = async () => {
  try {
    console.log("Generating QR codes...");
    await fs.remove(OUTPUT_FOLDER); // Remove the folder if it exists
    await fs.ensureDir(OUTPUT_FOLDER); // Recreate the folder

    const start_time = Date.now();

    for (let qrId = 1; qrId <= NUM_OF_QRS; qrId++) {
      await generateQRCode(qrId);
      if (qrId % PRINT_STATS_INTERVAL === 0 || qrId === NUM_OF_QRS) {
        const elapsed_time = ((Date.now() - start_time) / 1000).toFixed(2);
        console.log(
          `Generated ${qrId} QR codes. Elapsed time: ${elapsed_time} seconds.`
        );
        calculateStats();
      }
    }

    console.log("QR code generation completed.");
  } catch (error) {
    console.error("Error generating QR codes:", error);
  }
};

generateQRs();

File: old/simple_qrs.py
import os
import shutil
import time
import qrcode

NUM_OF_QRS = 1000000  # Change this to the desired number of QR codes (e.g., 1000000 for one million)

# Directory to save the QR code images
output_directory = "qr_codes"

# Remove the existing folder (if it exists) and recreate it
if os.path.exists(output_directory):
    shutil.rmtree(output_directory)

os.makedirs(output_directory)

# Generate and save QR codes
total_size = 0  # Total size of the folder
start_time = time.time()

for qr_id in range(1, NUM_OF_QRS + 1):
    qr = qrcode.QRCode(
        version=1,
        error_correction=qrcode.constants.ERROR_CORRECT_L,
        box_size=10,  # Adjust box size to control image size and readability
        border=4
    )
    qr.add_data(f"https://www.albertoqrtest.com/QR_{qr_id}")
    qr.make(fit=True)

    qr_image = qr.make_image(fill_color="black", back_color="white")

    # Save the QR code image with a filename corresponding to the QR_ID
    filename = f"qr_codes/QR_{qr_id}.png"
    qr_image.save(filename)

    # Calculate and accumulate the size of each QR code image
    total_size += os.path.getsize(filename)

end_time = time.time()

# Calculate average size of each QR code
average_size = total_size / NUM_OF_QRS

# Print stats to console
print(f"Generated {NUM_OF_QRS} QR codes.")
print(f"Total size of the folder: {total_size / (1024 * 1024):.2f} MB")
print(f"Average size of each QR code: {average_size / (1024):.2f} KB")
print(f"Time taken: {end_time - start_time:.2f} seconds")

# Write stats to a file
stats_file = "qr_stats.txt"
with open(stats_file, "w") as file:
    file.write(f"Generated {NUM_OF_QRS} QR codes.\n")
    file.write(f"Total size of the folder: {total_size / (1024 * 1024):.2f} MB\n")
    file.write(f"Average size of each QR code: {average_size / (1024):.2f} KB\n")
    file.write(f"Time taken: {end_time - start_time:.2f} seconds\n")

print(f"Stats saved to {stats_file}.")
